cpp/src/arrow/gpu/cuda_context.cc

```diff
@@ -165,6 +190,12 @@ class CudaDeviceManager::CudaDeviceManagerImpl {
     return (*out)->impl_->Init(devices_[device_number]);
   }
 
+  Status CreateSharedContext(int device_number, CUcontext ctx,
+                             std::shared_ptr<CudaContext>* out) {
+    *out = std::shared_ptr<CudaContext>(new CudaContext());
+    return (*out)->impl_->InitShared(devices_[device_number], ctx);
+  }
+
   Status GetContext(int device_number, std::shared_ptr<CudaContext>* out) {
     auto it = contexts_.find(device_number);
```

Comment:
Not very important for this PR, but I wonder if we still need the `contexts_` cache. On the one hand, it can avoid a memory allocation. On the other hand, it adds its own overhead (both memory & CPU). @wesm what was the reason for the cache?

Reply:
I would be OK with removing the cache. It might have been a premature over-design

Reply:
There is no noticeable impact to the timings of pyarrow cuda related tests when disabling or enabling `contexts_` cache. So, I'll remove `contexts_`.

Reply:
Removing `contexts_` also means that `CreateNewContext` can be removed in favor of `GetContext`.

Reply:
Cool